Restep testing

zachmprince opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason

Refs #30126

## Design

Tests can be run with `./run_tests --restep`. Or an input can be directly run with `moose-opt -i input.i --test-restep`.

The Transient executioner will then attempt to find the approximate half timestep of the simulation (if not it defaults to the first timestep). Once this timestep is reached, the timestep is "failed" in the executioner by manually setting `_last_solve_converged = false` in `TransientBase::takeStep()`, this will trigger a retry of the timestep, notably without any change in the timestep size. Supposedly, this should not change the answer of the simulation when this is done. This all follows very similar logic to `--recover` testing.

It seems most of the infrastructure in the framework is relatively impervious to this type of testing, but there were still many aspects that needed to be addressed. With the changes in this PR, all of the framework tests should be passing. The commits are pretty modular, describing the fix for various capabilities or a reasoning why restep wouldn't be possible/sensible for a given test. I intend on summarizing in this description in a follow-up edit.

## Impact

Better feature testing for new capabilities, ensuring robustness to failed solves.

## App patches

- [x] marmot (breaking): https://github.inl.gov/ncrc/marmot/pull/940
- [x] grizzly: https://github.inl.gov/ncrc/grizzly/pull/794
- [x] griffin: https://github.inl.gov/ncrc/griffin/pull/2532
- [x] TMAP8 (breaking): https://github.com/idaholab/TMAP8/pull/278
- [x] SAM (breaking): https://git-nse.egs.anl.gov/SAM/SAM/-/merge_requests/1086

loganharbour committed:
Start work on adding --restep testing capability

This adds the ability to systematically test
that the state when a timestep fails matches
the state at the beginning of the failed timestep

refs #30126

zachmprince committed:
Changing logic for determining 'mid' time step so that end_time, dt, and num_steps can all be set together

zachmprince committed:
Getting TimeSequenceSteppers working with restep

zachmprince committed:
PIDTransientControl was not handling failed solves quite right because it was looking at failed PP values

Re-golding necessary since test was showing bad behavior previously with failed solve

zachmprince committed:
Skipping a couple time integrators that can't handle failed solves

zachmprince committed:
Sample VPPs should be overwriting failed solves when complete history is requested

zachmprince committed:
Making sure to backup a reset or moved app in case a subsequent solve fails and requires restore

zachmprince committed:
Test object checking stateful old vpps needs to make sure itself isn't stateful

zachmprince committed:
Having time period tests run on timestep_begin makes more sense

zachmprince committed:
Don't want to advance random number generator in adaptivity test if solve didn't converge

zachmprince committed:
Fixing stateful test object since timestepSetup is called twice on the same timestep if solve failed

zachmprince committed:
Test has a BC dependent on the solution from the previous timestep cia a PP, restoring causes the PP to be ahead of the solution, which is fixed by adding timestep_begin to the execute_on

zachmprince committed:
Seeds aren't ever updated on timestep_end switching to timestep_begin allows restoration after a failed solve

zachmprince committed:
Making sure EXEC_FORCED is respected when redoing the timestep

This required a regolding of a test that previously had bad behaviour.

This multilevel multiapp test showed erroneous results in it's exodus out of the level 2 multiapp by two metrics:

1. There were only two timesteps in the exodus (the initial and final), despite the main app running 5 timesteps
2. Putting a printout in Exodus.C showed that the output was only happening on the first picard iteration, which is an unconverged solution.

This fix resulted in the test behaving as expected and the output had all the timesteps.

The second regold is for the same reasons since it uses the exact same mechanics.

zachmprince committed:
Fixing fixed point iteration for restep.

Several notes:
- Number of fixed point iterations can change when re-doing a timestep due to state retention of postprocessors, however the converged solution does not change. Therefore, many tests now ignore the CSV column containing the number of iterations.
- SecantSolve had a bug where it was using the solution from previous timestep for the update. This fix caused a small amount of regolding.
- Another bug was found in FixedPointSolve where _main_fixed_point_it was not being reset when the main app timestep failed, leading to out-of-sync data during the iteration update. Now MultiApp clears the solve state of the sub-apps' FixedPointSolve when a timestep fails, which now recognizes to reset _main_fixed_point_it.

zachmprince committed:
Handling repeated timesteps in PseudoTimestep properly

zachmprince committed:
RestartDiffusion needs to know when it is repeating a timestep

zachmprince committed:
SolutionTimeAdaptiveDT needs to recognize when a timestep is rejected

zachmprince committed:
Updating displaced mesh when multiapp solutions are restored, fixing transfers that are dependent on sub-app's displaced mesh

zachmprince committed:
Closing CSV file when it is done outputting.

Not closing was causing sub-apps to output duplicative timesteps on re-done timesteps and during fixed point iterations, hence the regolding of a few files that didn't make sense before.

zachmprince committed:
Removing unnecessary abort_on_solve_fail causing restep to exit early

zachmprince committed:
Removing statefulness of DynamicPointValueSampler

zachmprince committed:
LeastSqaresFitHistory needs to recognize when a timestep is repeated

zachmprince committed:
Skipping restep on tests that have abort_on_solve_fail=true

zachmprince committed:
Skipping restep of tests involving subdomain updaters

zachmprince committed:
Skipping tests that use stateful material property copy

zachmprince committed:
Skipping tests that are emulating restep

zachmprince committed:
Skipping tests related to solution invalidity

zachmprince committed:
Skipping tests that are doing a manual recover case

zachmprince committed:
Skipping various tests that I couldn't figure out how to fix

zachmprince committed:
Skipping various tests that are intended to be stateful

zachmprince committed:
Adding execute_on=multiapp_fixed_point_begin on objects executed on timestep_end where another object on timestep_begin depends on it.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2810727/), step Docs: sync website on e2de27a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30301/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2810742/), step Generate coverage on e2de27a wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>f54877</th>
      <th colspan=4>#30301 e2de27</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.30%</td>
      <td>85.31%</td>
      <td>+0.01%</td>
      <td>94.35%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>109952</td>
      <td>110033</td>
      <td>+81</td>
      <td>117</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18952</td>
      <td>18952</td>
      <td>-</td>
      <td>7</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30301/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30301/coverage/framework)


## Modules coverage

Inconsistent report tags were found between the head and base reports.
This can happen when reports are missing from either the head or the base.

Inconsistent tags:
  module-chemical_reactions-restep
  module-combined-restep
  module-contact-restep
  module-electromagnetics-restep
  module-external_petsc_solver-restep
  module-fluid_properties-restep
  module-fsi-restep
  module-functional_expansion_tools-restep
  module-geochemistry-restep
  module-heat_transfer-restep
  module-level_set-restep
  module-misc-restep
  module-navier_stokes-restep
  module-optimization-restep
  module-peridynamics-restep
  module-phase_field-restep
  module-porous_flow-restep
  module-ray_tracing-restep
  module-rdg-restep
  module-reactor-restep
  module-richards-restep
  module-scalar_transport-restep
  module-solid_mechanics-restep
  module-solid_properties-restep
  module-stochastic_tools-restep
  module-subchannel-restep
  module-thermal_hydraulics-restep
  module-xfem-restep

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30301/coverage/xfem)

</details>



This comment will be updated on new commits.

zachmprince committed:
Restep fixes for heat_transfer

- VPP is declaring vectors on every execute call, let's not do that
- Skipping test that is outputting on nonlinear
- Skipping tests that are recovering and skipping the restep timestep

zachmprince committed:
Fixing restep for solid_mechanics module

- Skipping tests doing their own restart
- Skipping tests that of 0 timesteps
- Skipping test that has plugin doing it's own output
- Skipping tests emulating explicit damage calculation
- Making controls execute on timestep_begin (when it makes sense to do that)
- Setting num_steps and/or dt in cli_args in tests where it is impossible for the executioner to predict the total number of timesteps a-priori
- Making sure dt isn't constrained when doing restep
- Getting AbaqusUserElement to swap stateful data when it is appropriate

zachmprince committed:
Restep fixes for functional_expansion_tools

- Removing fixed point iteration PP that changes when timestep is redone (required trivial regold)

zachmprince committed:
Restep fix for geochemistry module

- Test was emulating an custom action, but was not quite the same as the action-using equivalent.

zachmprince committed:
Restep fixes for misc module

- Sensor postprocessors need to detect when a timestep is being repeated

zachmprince committed:
Restep fix for optimization module

- Makings sure simulation values in OptimizationData aren't accumulated if a timestep is repeated

zachmprince committed:
Restep fix for peridynamics module

- Small change in logic in FixedPointSolve with custom PP convergence to use old PP value for comparison on first iteration.

zachmprince committed:
Restep fixes for rdg module

- Skipping tests using ExplicitMidpoint

zachmprince committed:
Restep fixes for the richards modules

- Using num_steps in some tests so restep can properly guess the middle timestep

zachmprince committed:
Restep fix for scalar_transport module

- Steady-state detection test needed an appoximate number of steps so restep can happen before steady-state

zachmprince committed:
Restep fix for solid_properties module

- Skipping test that has abort_on_solve_fail = true

zachmprince committed:
Restep fixes for navier_stokes module

- Skipping tests that have abort_on_solve_fail = true
- Setting num_steps for tests that use IterationAdaptiveDT
- Setting num_steps for tests that use steady-state detection
- Adjusting tolerance for test that was very close to the gold

zachmprince committed:
Some restep fixes for phase_field module

- Skipping tests that are doing recover
- Skipping tests that don't have any timesteps
- Objects using RandomData need to register the RNG update on timestep_begin since updateSeeds is not called on timestep_end
- Using _t_step_old logic in a couple objects to recover state from before the failed timestep

zachmprince committed:
Restep fixes for porous_flow module

- Skipping tests that are doing recovery
- Some tests had function evaluations that were delayed due to their execution on timestep_end, switching to timestep_begin and modifying the time sequence fixed the issue. Very minor regolding was required, but I believe the changes are more faithful to the test than what was there before.
- Some tests used timestep_begin to compute the "old" value of a PP, which doesn't work when a timestep is repeated. I modified the test such that every PP can be executed on timestep_end without statefulness.

zachmprince committed:
Partial restep fixes for contact module

- Skipping tests that have abort_on_solve_fail=true
- Skipping tests that don't have any timesteps
- Skipping a tests that outputs on nonlinear
- Changing a initial adaptivity test Adaptivity/steps to cycles_per_step to prevent running adaptivity after initial time steps, requiring a regolding. Related to #30390

zachmprince committed:
Skipping restep for many tests in thermal_hydraulics module

- Tests have zero timesteps
- Tests have abort_on_solve_fail=true

zachmprince committed:
Restep updates from devel rebase:

    - Complication in point sampler complete history when old value is requested due to new postprocessor restore
    - New solid_mechanics time integrator doesn't work with restep
    - New test to skip in THM

zachmprince committed:
Restep skips and num_step fixes in combined module

zachmprince committed:
Regolding heavy THM tests

These tests have a failed solve in them, which the BDF2 time integrator wasn't handling properly due to the erroneous update to _dt_old. The is error was fixed when going through the restep fixes in the solid_mechanics module.

loganharbour commented:
@zachmprince let me know when this is ready for review

zachmprince commented:
@loganharbour This is ready for review

- Currently failing module tests. I can't figure out why these are failing and would need to outsource debugging. Will need to decide whether to pursue fixing or at least explaining, or simply skipping restep for these tests.
    ```
    xfem/test:solid_mechanics_basic.crack_propagation_ave ................................................................................................................ FAILED (EXODIFF)
    contact/test:pdass_problems.cylinder_friction_penalty_al_frictional_action_amg_bussetta_simple ....................................................................... FAILED (CSVDIFF)
    xfem/test:mechanical_constraint.glued_penalty ........................................................................................................................ FAILED (EXODIFF)
    xfem/test:solid_mechanics_basic.crack_propagation_single_point ....................................................................................................... FAILED (EXODIFF)
    contact/test:mortar_aux_kernels.block-dynamics-aux-wear .............................................................................................................. FAILED (EXODIFF)
    phase_field/test:grain_tracker_test.changing_avg_volume .............................................................................................................. FAILED (CSVDIFF)
    combined/test:gap_heat_transfer_jac.gap-heat-transfer-jac ............................................................................................................. FAILED (ERRMSG)
    ```
- Grizzly failures addressed in https://github.inl.gov/ncrc/grizzly/pull/794
- Need help with Marmot since I don't have access
- Griffin failures addressed in https://github.inl.gov/ncrc/griffin/pull/2532
- SAM failures simply require a regold, but I don't have access to ANL repo.

loganharbour commented on framework/src/executioners/TransientBase.C:
```diff
@@ -428,6 +482,29 @@ TransientBase::takeStep(Real input_dt)
+    // We shouldn't try to test resolving if the last solve was
+    // a failure. This might actually not be needed, but I'm
+    // not confident
+    // if (!lastSolveConverged())
```
Should this remain?

zachmprince replied:
I guess not. I'll remove it.

loganharbour commented on framework/src/vectorpostprocessors/SamplerBase.C:
```diff
@@ -41,6 +44,10 @@ SamplerBase::SamplerBase(const InputParameters & parameters,
   : _sampler_params(parameters),
     _vpp(vpp),
     _comm(comm),
+    _sampler_transient(dynamic_cast<TransientBase *>(
```
Can this cast fail?

zachmprince replied:
Yes, but it should just return a nullptr, right? That's checked before calling it below.

zachmprince committed:
Some undefined behavior was apparently happening with the fixed point solver, simple fix, though

zachmprince committed:
Removing comment with error for a timestep failing at the point of restep

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit 9bc8c929b3d21601b4d73b4fb2eb02ed06aba657
+Subproject commit fdf470ed8e7ee0cd5b65a624e2e136ed1f26c05f
```
**Caution!** This contains a submodule update

zachmprince committed:
More tests that have abort_on_solve_fail=true in thermal hydraulics module

zachmprince committed:
Skipping recover tests in contact that are only doing 1 timestep

loganharbour commented on framework/include/executioners/SecantSolve.h:
```diff
@@ -79,27 +79,30 @@ class SecantSolve : public FixedPointSolve
   virtual void printFixedPointConvergenceHistory() override final;
 
+  /// Default TagID if storage is not allocated
+  static TagID _default_tag_id;
```
Few things:

- For static variables, don't prefix with an underscore. Underscore implies object state, even though this doesn't change
- These should typically be const
- While you can initialize these in the header, we typically prefer to do this in the initializer list

Fortunately - `INVALID_TAG_ID` already exists in `MooseTypes.h` and this should be used instead. So, remove this definition and declaration, use `INVALID_TAG_ID`, and set it in the initializer list instead.

zachmprince committed:
Extending usage of INVALID_TAG_ID and check to other fixed point solvers

loganharbour commented:
Go ahead and do the rebase - I've got everything as viewed and all looks well.

zachmprince committed:
Mentioning new issue for skipping tests related to mesh modifiers

GiudGiud commented on framework/src/executioners/SecantSolve.C:
```diff
@@ -105,10 +115,17 @@ SecantSolve::saveVariableValues(const bool primary)
+    fxn_m2 = solution;
+  // Otherwise we just advance
+  else
+    fxn_m2 = fxn_m1;
```
thanks for catching that!

GiudGiud commented on framework/src/postprocessors/PseudoTimestep.C:
```diff
@@ -171,11 +171,23 @@ PseudoTimestep::execute()
   // at the end of each timestep call the postprocessor to set values for dt
   if (_current_execute_flag == EXEC_TIMESTEP_END)
   {
+    // This is all incase a timestep fails and needs to be re-done
```
```suggestion
    // This is all in case a timestep fails and needs to be re-done
```

GiudGiud reviewed (changes requested):
Absolutely huge PR hitting the whole repo. Thanks for taking care of that, it's an insane amount of work to cover all this. 

More comments:
please add explanation & advice on how to support restepping to doco

in the test objects, it seems that any statefulness needs to be removed by handling the restep explicitly (use a t_old). Why is that? I would hope we can use restartableData instead to get the previous value back

Another check we should do is restep x2. So restepping 2x in a row. This happens in simulations quite often. This would check if the logic added only works once or is robust to this

loganharbour commented on test/src/kernels/RestartDiffusion.C:
```diff
@@ -22,14 +22,20 @@ RestartDiffusion::validParams()
   : Kernel(parameters),
     _coef(getParam<Real>("coef")),
-    _current_coef(declareRestartableData<Real>("current_coef", 1))
+    _current_coef(declareRestartableData<Real>("current_coef", 1)),
```
still no

zachmprince commented on framework/include/executioners/FixedPointSolve.h:
```diff
@@ -245,6 +245,7 @@ class FixedPointSolve : public SolveObject
 private:
   /// Postprocessor value for user-defined fixed point convergence check
   const PostprocessorValue * const _fixed_point_custom_pp;
+  const PostprocessorValue * const _fixed_point_custom_pp_old;
```
No, `_pp_old` is the value from the previous fixed-point iteration; `_fixed_point_custom_pp_old` is the value from the previous timestep. We want to make sure we set `_pp_old` to the previous timestep on the first iteration after the timestep fails.

GiudGiud replied:
the attributes needs to be renamed to make that clear
and we need a docstring

zachmprince replied:
New variables and docstrings implemented

GiudGiud commented on framework/include/base/MooseApp.h:
```diff
@@ -545,11 +545,16 @@ class MooseApp : public ConsoleStreamInterface,
   bool testCheckpointHalfTransient() const { return _test_checkpoint_half_transient; }
 
+  /**
+   * Whether or not this simulation should fail its middle timestep and repeat (for testing)
```
A restep needs to work on all steps, with debatedly the exception of the first one
If a test does not work with restep on all steps then either:
- it s due to the test design and we can skip it
- we add an error in the object that does not support restep and we can skip it

GiudGiud replied:
noice logic seems even more robust than I was thinking

zachmprince commented on modules/richards/test/tests/gravity_head_1/gh20.i:
```diff
@@ -116,7 +116,7 @@
   type = Transient
   solve_type = Newton
-  dt = 1E0
+  num_steps = 25
```
The number of steps is fixed for this test, but we can't programmatically determine it because that function could depend on things like postprocessors. I think is a viable technique to get restep working. Logan and I believe restepping the middle timestep is more robust since it generally allows the system to evolve to their inflection point.

zachmprince replied:
This change is no longer needed

zachmprince commented on test/tests/multiapps/move_and_reset/tests:
```diff
@@ -19,6 +19,15 @@
+      # causes the exodus object to think it needs to increment again. This
+      # doesn't cause issues without restepping because it doesn't take into
+      # account the mesh change on the subsequent output call, but it does if we
+      # are re-doing that timestep.
```
No, for three reasons: 1) this is a very improbable series of events (move and reset app with its own sub-app + failed timestep at the same time), 2) the logic would be quite complicated to catch, and 3) the result is still valid, it will just have different file names than if the timestep didn't repeat.

GiudGiud replied:
3) is a good reason!

zachmprince commented on test/tests/materials/stateful_prop/tests:
```diff
@@ -100,6 +100,9 @@
     max_parallel = 1
+    # There currently isn't a way to undo a stateful material copy when
+    # trying to re-do the timestep where it happened
+    restep = false
```
No, I think it would be up to the object calling `MaterialData::copy` to throw this error. This test is just using a test object calling that API.

GiudGiud commented on modules/porous_flow/test/tests/sinks/s06.i:
```diff
@@ -172,25 +133,37 @@
   [m10]
-    type = FunctionValuePostprocessor
-    function = mass10
+    type = ParsedPostprocessor
```
This would not really show up on a search for someone looking at their test re-stepping problem. 
I would recommend a new one linking to this old one

zachmprince replied:
I don't think that type of issue could ever be resolved, though. If there are dependency issues in any form, restep might show a failure. The real fix is to resolve the dependency issue like I did here. I will be adding documentation on the restep capability per your suggestion, what if I mention this as a common cause of failure there?

GiudGiud replied:
>I don't think that type of issue could ever be resolved, though. 

you're not wrong...

>  The real fix is to resolve the dependency issue like I did here.

The problem is that every future developer trying to get restepping to pass will have to do the same thing as you. It's not ideal.

> what if I mention this as a common cause of failure there?

definitely a good step forward!

zachmprince replied:
https://github.com/idaholab/moose/issues/30996

zachmprince committed:
Addressing Guillaume's easy comments from his first restep review

zachmprince committed:
Reverting some of the changes from the porous flow restep fixes

- Postprocessor restore on failed timestep fixes the previous issues seen
- Keeping the dependency resolution fixes since they are good practice
- Recover tests still need to be skipped

zachmprince committed:
Adding informational message in controls if testing restep and control is executed on timestep_end

zachmprince committed:
Using EXEC_TRANSFER for user objects used in transfers

zachmprince committed:
Adding error when restoring solution and not zeroed vectors are requested

zachmprince committed:
Reverting regolding in functional expansion tools and using custom_cmp instead

zachmprince committed:
Changing logic in restep timestep search to also include end_time and take into account pseudo-transient

- Reverts changes to tests that forcefully set num_steps

zachmprince committed:
Restep tests and documentation

zachmprince commented on modules/porous_flow/test/tests/hysteresis/1phase.i:
```diff
@@ -111,7 +111,8 @@
   [flux]
     type = FunctionValuePostprocessor
-    function = 'if(t <= 9, -10, 10)'
+    function = 'if(t <= 0.5, 0.0, if(t <= 9.5, -10, 10))'
```
No longer needed

GiudGiud commented on modules/solid_mechanics/src/userobjects/AbaqusUserElement.C:
```diff
@@ -26,7 +26,7 @@ AbaqusUserElement::validParams()
   ExecFlagEnum & exec_enum = params.set<ExecFlagEnum>("execute_on", true);
   exec_enum.addAvailableFlags(EXEC_PRE_KERNELS);
-  exec_enum = {EXEC_PRE_KERNELS, EXEC_TIMESTEP_END};
+  exec_enum = {EXEC_PRE_KERNELS};
```
ok works.

GiudGiud commented on test/tests/controls/time_periods/kernels/kernels.i:
```diff
@@ -61,6 +61,7 @@
     disable_objects = '*::diff1'
     start_time = '0'
-    end_time = '0.49'
+    end_time = '0.51'
+    execute_on = 'timestep_begin'
```
works for me

GiudGiud commented on python/doc/content/python/TestHarness.md:
```diff
@@ -45,6 +45,32 @@ through, but some of the important ones are:
+    [Predictor::timestepSetup](framework/src/predictors/Predictor.C) for an
+    example.
+  - Specifying `Executioner/abort_on_solve_fail=true` in the test input or
+    specification `cli_args`.
```
```suggestion
    in the command line arguments (`cli_args`).
```

GiudGiud commented on test/tests/multiapps/picard_postprocessor/tests:
```diff
@@ -30,6 +30,7 @@
       input = 'transient_main.i'
       csvdiff = 'transient_main_out.csv transient_main_out_sub0_csv.csv'
       cli_args = 'Executioner/fixed_point_max_its=5 Executioner/fixed_point_abs_tol=1e-10'
+      restep = false # Test will be repeating it's own time steps
```
```suggestion
      restep = false # Test will be repeating its own time steps
```

GiudGiud reviewed (commented):
last round

zachmprince commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -845,6 +857,16 @@ MultiApp::restore(bool force)
     }
+
+    // Make sure the displaced mesh on the multiapp is up-to-date with displacement variables
+    for (const auto & app_ptr : _apps)
```
I'm honestly not sure what the state of the displaced mesh is when it is restored, it could very well be undisplaced. I think it is probably safer to just displace the mesh even if we are keeping solutions. Especially since displacing the mesh is not a heavy operation.

zachmprince committed:
Addressing final round of comments for restep

zachmprince committed:
Changing design page of restep tests because the test/doc config doesn't include TestHarness, and neither do a lot of dependent apps

zachmprince commented:
> test failures might be new tests?

Those restep tests failure are some that I could not figure out why they weren't working, nor could I justify having `restep = false`. They will be optionally triggered tests for now until we can recruit experts to debug them.

zachmprince committed:
Skipping remaining failing tests to get modules passing see #31054

zachmprince committed:
Fixing valgrind failure caused by uninitialized variable in SteadyBase

zachmprince committed:
Regolding thermal hydraulics heavy test that has the proliferant BDF2 bug fix

YaqiWang commented:
This seems introducing a valgrind issue in Griffin with the custom Richardson executioner. I guess the fix is to set the default for `_last_solve_converged`. Why do we have to this now?
